File: web-ui/types/graph.ts
export type GraphMetadata = {
  graph_id: string;
  owner_username: string;
  created_at: Date;
  updated_at: Date;
  name: string;
  description: string;
  user_role: Role;
  is_public: boolean;
  is_bookmarked_by_user: boolean;
  is_cheered_by_user: boolean;
  nb_data_nodes: number;
  nb_data_edges: number;
  nb_cheers: number;
  nb_bookmarks: number;
};

export enum Role {
  OWNER = "owner",
  ADMIN = "admin",
  EDITOR = "editor",
  VIEWER = "viewer",
  NONE = "none"
}

export type GraphSchema = {
  nodes: NodeSchema[];
  edges: EdgeSchema[];
};

export type NodeSchema = {
  node_id: string;
  label: string; // Must be unique per graph (including edge labels)
  formated_label: string; // Must be unique per graph (including edge labels)
  color: string;
  properties: Property[];
};

export type EdgeSchema = {
  edge_id: string;
  label: string; // Must be unique per graph (including node labels)
  formated_label: string; // Must be unique per graph (including node labels)
  color: string;
  properties: Property[];
};

export type Property = {
  property_id: string;
  name: string; // ex: "nombre d'enfants" (must be uniques per node/edge)
  formated_name: string; // ex: "nombre_d_enfants" (must be uniques per node/edge)
  metadata: {
    property_type: PropertyType;
    details: {
      min: number | null;
      max: number | null;
      options: string[] | null;
      required: boolean;
      default_value: PropertyValue | null;
    };
  };
};

export enum PropertyType {
  INTEGER = "integer",
  FLOAT = "float",
  STRING = "string",
  BOOLEAN = "boolean",
  DATE = "date",
  TIME = "time",
  RANGE = "range",
  SELECT = "select",
  MULTISELECT = "multiselect"
}

export type PropertyValue = string | number | boolean;

export type GraphData = {
  nodes: NodeData[];
  edges: EdgeData[];
};

export type NodeData = {
  graph_id: string;
  node_id: string;
  label: string;
  properties: Record<string, PropertyValue>;
};

export type EdgeData = {
  graph_id: string;
  edge_id: string;
  from_id: string;
  to_id: string; // Couple (from_id to_id) with label must be unique per graph
  label: string;
  properties: Record<string, PropertyValue>;
};

export type ProcessedGraphData = {
  nodes: ProcessedNodeData[];
  links: ProcessedEdgeData[];
};

export type ProcessedNodeData = {
  id: string;
  label: string;
  color: string;
};

export type ProcessedEdgeData = {
  source: string;
  target: string;
  label: string;
  color: string;
};

File: metadata/src/dtos/responses/graph.rs
use crate::{
    dtos::{EdgeDataResponse, EdgeSchemaResponse, NodeDataResponse, NodeSchemaResponse, RoleDto},
    models::{GraphData, GraphId, GraphMetadata, GraphSchema, Reddit},
};
use chrono::{DateTime, Utc};
use serde::Serialize;

#[derive(Debug, Serialize)]
pub struct RedditDto {}

impl From<Reddit> for RedditDto {
    fn from(_reddit: Reddit) -> Self {
        RedditDto {}
    }
}

#[derive(Debug, Serialize)]
pub struct GraphMetadataResponse {
    pub graph_id: GraphId,
    pub name: String,
    pub description: String,
    pub is_public: bool,
    pub reddit: RedditDto,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub nb_data_nodes: u32,
    pub nb_data_edges: u32,
    pub owner_username: String,
    pub user_role: RoleDto,
    pub is_bookmarked_by_user: bool,
    pub is_cheered_by_user: bool,
    pub nb_bookmarks: u32,
    pub nb_cheers: u32,
}

impl From<GraphMetadata> for GraphMetadataResponse {
    fn from(metadata: GraphMetadata) -> Self {
        Self {
            graph_id: metadata.graph_id,
            name: metadata.name,
            description: metadata.description,
            is_public: metadata.is_public,
            reddit: metadata.reddit.into(),
            created_at: metadata.created_at,
            updated_at: metadata.updated_at,
            nb_data_nodes: metadata.nb_data_nodes,
            nb_data_edges: metadata.nb_data_edges,
            owner_username: metadata.owner_username,
            user_role: metadata.user_role.into(),
            is_bookmarked_by_user: metadata.is_bookmarked_by_user,
            is_cheered_by_user: metadata.is_cheered_by_user,
            nb_bookmarks: metadata.nb_bookmarks,
            nb_cheers: metadata.nb_cheers,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct GraphSchemaResponse {
    pub nodes: Vec<NodeSchemaResponse>,
    pub edges: Vec<EdgeSchemaResponse>,
}

impl From<GraphSchema> for GraphSchemaResponse {
    fn from(schema: GraphSchema) -> Self {
        Self {
            nodes: schema
                .nodes
                .into_iter()
                .map(NodeSchemaResponse::from)
                .collect(),
            edges: schema
                .edges
                .into_iter()
                .map(EdgeSchemaResponse::from)
                .collect(),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct GraphDataResponse {
    pub nodes: Vec<NodeDataResponse>,
    pub edges: Vec<EdgeDataResponse>,
}

impl From<GraphData> for GraphDataResponse {
    fn from(data: GraphData) -> Self {
        Self {
            nodes: data.nodes.into_iter().map(NodeDataResponse::from).collect(),
            edges: data.edges.into_iter().map(EdgeDataResponse::from).collect(),
        }
    }
}
